src/blocked.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Redis Ltd. 2023 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#include "redisraft.h"
+
+#include <string.h>
+
+/* A list of all current blocked commands */
+
+BlockedCommand *addBlockedCommand(raft_index_t idx, unsigned long long session, const char *data, size_t data_len, RaftReq *req, RedisModuleCallReply *reply)
+{
+    BlockedCommand *bc = RedisModule_Calloc(1, sizeof(BlockedCommand));
+    sc_list_init(&bc->blocked_list);
+    sc_list_add_tail(&redis_raft.blocked_command_list, &bc->blocked_list);
+
+    bc->idx = idx;
+    bc->session = session;
+    bc->data = RedisModule_Alloc(data_len);
+    memcpy(bc->data, data, data_len);
+    bc->data_len = data_len;
+    bc->req = req;
+
+    RedisModule_DictSetC(redis_raft.blocked_command_dict, &bc->idx, sizeof(bc->idx), bc);
+
+    return bc;
+}
+
+BlockedCommand *getAndDeleteBlockedCommand(raft_index_t idx)
+{
+    BlockedCommand *blocked = NULL;
+    RedisModule_DictDelC(redis_raft.blocked_command_dict, &idx, sizeof(idx), &blocked);
+    if (blocked == NULL) {
+        return NULL;
+    }
+
+    sc_list_del(&redis_raft.blocked_command_list, &blocked->blocked_list);
+
+    return blocked;
+}
+
+
+void freeBlockedCommand(BlockedCommand *bc)
+{
+    if (bc->data) {
+        RedisModule_Free(bc->data);
+        bc->data = NULL;
+    }
+    RedisModule_Free(bc);
```

Comment:
```suggestion
    RedisModule_Free(bc->data);
    RedisModule_Free(bc);
```
You can avoid null check if you want.

Reply:
I think I'd rather have it in case in future its possible to have a BlockedCommand struct with a null data element